[DOC] Example comparing spectral connectivity computed over time or over trials

Avoide opened the pull request:
PR Description
--------------

This PR adds an example comparing spectral connectivity over time or trials and closes #70 

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-connectivity/blob/main/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

Avoide commented:
Hi @adam2392, here is a draft for an example comparing spectral connectivity computed over time or over trials on simulated data and sample EEG data.

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,392 @@
+# In this example, we focus on the two functions
+# :func:`mne_connectivity.spectral_connectivity_epochs` and
+# :func:`mne_connectivity.spectral_connectivity_time`,
+# which both computes connectivity from epoched data :class:`mne.Epochs`.
+#
+# Both functions contain the options to choose the connectivity measurements
+# of interest and the method to decompose the frequency content.
+# The crucial difference between when to use which function lies on
+# the experimental setup and type of tasks.
+#
+# If the data is obtained for repeated trials of a time-locked event,
+# e.g. ERP data, then :func:`mne_connectivity.spectral_connectivity_epochs`
+# is most likely the corresponding function to utilize.
+#
+# If the data is obtained from resting-state or free-moving tasks,
+# e.g. a mirror-game paradigm, then
+# :func:`mne_connectivity.spectral_connectivity_time` might be better suited.
+#
+# The way connectivity is computed for the two functions are slightly different,
+# thus their interpretations and the hypotheses being tested are also different.
+#
+# Connectivity over trials, as computed by
+# :func:`mne_connectivity.spectral_connectivity_epochs` assume epochs have
+# been created around each trial and the function estimates the connectivity
+# at each timepoint over all the repeated trials for the same event.
+# This gives a high temporal resolution, which is often desired for ERP
+# analysis where you are interested in when an effect occurs.
+# However, this approach is not feasible for single-trial data and will
+# result in errorful values when computed on one epoch.
+#
+# On the other hand, if you are interested in determining connectivity over
+# time for single-trials or from experiments that do not involve exactly
+# repeated trials, then :func:`mne_connectivity.spectral_connectivity_time`
+# should be employed. This function also takes data in the form of
+# :class:`mne.Epochs`, but it may consist of a single epoch.
+# If there are multiple epochs, the connectivity over time is computed for
+# epoch separately, with the option to average over epochs.
+#
+# To better illustrate the differences between the two functions,
+# we will employ them on two simulated cases.
```
Wdyt about adding this to the top section above the imports? E.g. right after

"A brief background on the difference between the two conditions is provided,
 followed by examples on simulated data and real EEG data."

Avoide replied:
Sure, I'll move it to the top

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,392 @@
+#
+# Let's generate some simulated data in the format of :class:`mne.EpochsArray`.
+# In this case, we will use random data for 3 channels over 5 epochs, but
+# all the epochs are just exact replicates of the first epoch.
```
```suggestion
# all the epochs are just exact replicates of the first epoch. This simulates when data is collected over
# an event of interest where we **assume** the connectivity structure is the same over each event.
```

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,392 @@
+# In this case, we will use random data for 3 channels over 5 epochs, but
+# all the epochs are just exact replicates of the first epoch.
+
+np.random.seed(1234) # set seed for reproducibility
```
I would set this at the very top. And instead of setting a seed, I would use a random number generator, which is considered more stable (https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html).

You can do this with `rng = np.default_rng(1234)` and then everytime you have `np.random.*`, you replace it with `rng.*` 

See: https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.default_rng

adam2392 reviewed (changes requested):
Some style failures on the CI. Do you mind fixing those? You can run `make pep` to check these locally.

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,392 @@
+# :class:`mne_connectivity.SpectralConnectivity`, which does not have
+# single timepoint resolution.
+
+con_epochs_array = np.mean(con_epochs_array, axis=4) # average over timepoints
```
If we want to average `SpectroTemporalConnectivity` container over time, can we use the `combine` method? https://mne.tools/mne-connectivity/dev/generated/mne_connectivity.SpectroTemporalConnectivity.html#mne_connectivity.SpectroTemporalConnectivity.combine

Avoide replied:
I just tried, but combine only works for averaging over epochs, which ``spectral_connectivity_epochs`` already do if I understood the ``nave`` correctly (when printing SpectroTemporalConnectivity).

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,392 @@
+    print(f"Connectivity method: {connectivity_methods[c]}\n"+
+          f"{con_time_array[c,:,:,foi]}")
+
+###############################################################################
+# We see that the connectivity over time are not 1, since the signals were
+# randomly generated and therefore the phase differences between channels
+# are also random over time.
+#
+# **Case 2: 10 Hz sinus waves with different phases.**
```
Perhaps break up the sectioning of "conclusion" of the previous case with the start of "Case 2"?

Avoide commented:
@adam2392 I should have fixed the code style errors and have updated the places you mentioned in the review in the new commit

adam2392 reviewed (commented):
There are some errors on the CI still: https://github.com/mne-tools/mne-connectivity/actions/runs/4322408221/jobs/7573270726. You can check these locally running `make pep`.

Hmm... for some reason the CI build docs is not working. Can you investigate? Does it build properly on your local?

Avoide commented:
> There are some errors on the CI still: https://github.com/mne-tools/mne-connectivity/actions/runs/4322408221/jobs/7573270726. You can check these locally running `make pep`.
> 
> Hmm... for some reason the CI build docs is not working. Can you investigate? Does it build properly on your local?

hmm I didn't see those spelling errors when I ran ``make pep`` locally. But I just fixed them according to the error on the CI job in the newest commit. Hopefully that should work.

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,401 @@
+###############################################################################
+# Real data demonstration
+# -----------------------
+# To finish this example, we will compute connectivity for a sample EEG data.
```
Can you add some details on what we expect to see and the motivation for picking different analysis options?

E.g. we focus on theta power. Why? What do we expect to see? At the timepoint of highest-connectivity, what does that tell us?

Avoide replied:
I haven't worked with ERP analysis myself, so I'm not really sure what to expect.
I talked with a colleague and she said you usually expect some theta power differences in the occipital cortex for a visual task, but wasn't sure regarding the expected connectivity, so I purposefully left it blank since I'm also not confident in the "ground truth" in the example.

adam2392 replied:
Perhaps what we can say is something along those lines (ideally with a reference perhaps)?

E.g.

```
Since the sample data is a visual task, we should generally expect to see some theta power
ERP(?) differences in the occipital cortex :footcite:`<insert_ref see references.bib and other files>`.
Here, we will therefore analyze the correlations in the frequency domain, centered at the theta band.
```

Avoide replied:
I found this [reference](https://www.sciencedirect.com/science/article/pii/S0926641004000308?via%3Dihub) about how phase locked alpha and theta oscillations are related to P1 and N1 and tried to relate it to the results in the example.

adam2392 reviewed (changes requested):
Looking good! Leaving some more in-depth review comments now.

For now, the cirlceCI is failing due to some upstream issues in mne-python.

So I'm leaving some additional comments based on rendering the docs locally. Is it possible to add colorbars to the plots?

Avoide commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,401 @@
+con_epochs_array = con_epochs.get_data(output="dense")
+
+fig = plt.figure()
+im = plt.imshow(con_epochs_array[:, :, 0, t_con_max])
+fig.colorbar(im)
+plt.ylabel("Channel 1")
+plt.xlabel("Channel 2")
```
wdyt of Channel X and Channel Y, or just have Channel and Channel without any designation? I think source might be confusing as it could refer to source localized channel.

adam2392 replied:
Sure, maybe `Channels` and `Channels` to denote that these are channels x channels connectivity?

Avoide commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,401 @@
+fmin = tuple([list(Freq_Bands.values())[f][0] for f in range(len(Freq_Bands))])
+fmax = tuple([list(Freq_Bands.values())[f][1] for f in range(len(Freq_Bands))])
+
+# We specify the connectivity measurements
+connectivity_methods = ["plv", "wpli"]
+n_con_methods = len(connectivity_methods)
```
The choice was actually because I implemented my own plv and wpli functions over time a few years ago, as MNE-connectivity didn't have the functionality, and wanted to test it up against it (Issue #17)

But I have now changed it to use two more different measurements (coh and plv) as that might make more sense.

adam2392 commented on requirements_doc.txt:
```diff
@@ -12,4 +12,5 @@ typing-extensions
-pooch
\ No newline at end of file
+pooch
+circleci
```
Should be mne-qt-browser

Avoide commented:
I updated the example, and also added circleci to the requirement_doc.txt.
There was also a small typo in contributing.md (requirement_doc.text, which I fixed to requirement_doc.txt).

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,409 @@
+plt.xlabel("Time (s)")
+plt.ylabel("Global theta wPLI over trials")
+
+# Get the timepoint with highest global connectivity
+t_con_max = np.argmax(global_con_epochs)
+
+# Plot the connectivity matrix at the timepoint with highest global wPLI
+con_epochs_array = con_epochs.get_data(output="dense")
+
```
Perhaps in line w/ https://github.com/mne-tools/mne-connectivity/pull/129/files#r1128169084, we can then add some comments about how the highest time point connectivity means something perhaps?

Is it correct to say that we are "extracting the timepoint at which there is the most correlation in the theta frequency band as measured by wPLI (or some other connectivity metric) and visualizing the connectivity"?

Maybe we can also mention something about the output of the final connectivity matrix. I.e. are the channels in occipital cortex the ones with generally higher correlation?

Avoide replied:
I added a bit in the conclusion and also the plotted the sensor connectivity, to visualize the 20 highest connections which was between the occipital and frontal areas.

adam2392 commented:
https://output.circle-artifacts.com/output/job/21c46aa5-0fb7-4326-8cdf-bb256e43ca8a/artifacts/0/dev/auto_examples/compare_connectivity_over_time_over_trial.html#sphx-glr-auto-examples-compare-connectivity-over-time-over-trial-py

- can you add a label to the colorbars?
- looks like high connectivity at 0.15-0.2 seconds after the visual task. Are there any references for this?

adam2392 commented:
Rendered docs: https://output.circle-artifacts.com/output/job/03472a75-4383-47d6-923f-9364a9a1b443/artifacts/0/dev/index.html

adam2392 commented:
So it seems that circleCI always blocks your docs build because perhaps you haven't connected your GH account to there :(. Any chance you can just create an account and then the docs build should proceed without someone in the future explicitly running it? (Annoying feature, but I guess there for security) Xref: https://support.circleci.com/hc/en-us/articles/360037744473-What-is-an-Unregistered-User-

Can you also add the colorbar title?

Avoide commented:
Sure, I just registered my GH at circleCI. Let's see if it works.

I also added the label "Connectivity" for the colorbars.

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,427 @@
+# Additionally, there are also multiple methods to estimate the frequency
+# content, e.g. Fourier transforms, Morlet wavelets and multitapers.
+#
+# In this example, we focus on the two functions
+# :func:`mne_connectivity.spectral_connectivity_epochs` and
+# :func:`mne_connectivity.spectral_connectivity_time`,
+# which both computes connectivity from epoched data :class:`mne.Epochs`.
```
```suggestion
# In this example, we focus on the two functions, which both computes connectivity from epoched data :class:`mne.Epochs`:
# - :func:`mne_connectivity.spectral_connectivity_epochs` and
# - :func:`mne_connectivity.spectral_connectivity_time`
```

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,427 @@
+# the experimental setup and type of tasks.
+#
+# If the data is obtained for repeated trials of a time-locked event,
+# e.g. ERP data, then :func:`mne_connectivity.spectral_connectivity_epochs`
```
Not everyone that I know of that works w/ mne is familiar w/ ERP data. Perhaps we can link either another MNE-python tutorial/example/glossary, or a paper that defines it.

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,427 @@
+# is most likely the corresponding function to utilize.
+#
+# If the data is obtained from resting-state or free-moving tasks,
+# e.g. a mirror-game paradigm, then
```
Same comment here, where ideally we give a link to help define this.

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,427 @@
+# If the data is obtained from resting-state or free-moving tasks,
+# e.g. a mirror-game paradigm, then
+# :func:`mne_connectivity.spectral_connectivity_time` might be better suited.
+#
```
```suggestion
#
# **Assumptions and Interpretations**
# 
```

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,427 @@
+# epoch separately, with the option to average over epochs.
+#
+# To better illustrate the differences between the two functions,
+# we will employ them on two simulated cases.
```
```suggestion
# we will employ them on two simulated cases and also analyze a real visual-task dataset.
```

adam2392 commented on examples/compare_connectivity_over_time_over_trial.py:
```diff
@@ -0,0 +1,427 @@
+# Both functions serve their specific roles, and it's important to use the
+# correct function for the corresponding task to interpret the analysis.
+#
+# We also briefly analyzed a visual task EEG sample, where we found that
```
```suggestion
# We also briefly analyzed a visual task EEG sample using the :func:`mne_connectivity.spectral_connectivity_epochs`, where we found that
```

adam2392 reviewed (commented):
Some final nitpicks and then this LGTM for merge.

Some open-ended comments, so lmk if you disagree.

adam2392 commented:
If a comment is resolved/addressed, you can feel free to click the "Resolve conversation"

Avoide commented:
Thanks for the all comments. And the practical instructions are also useful to know, as this is my first PR.

I have incorporated all your newest comments in the latest commit.

adam2392 reviewed (approved):
LGTM

adam2392 commented:
Thanks @Avoide!
